text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,879 @@
+functions closer to the ones found in other languages, as well as closer to
+synchronous functions.
+
+The `CancellationToken/StopToken` approach, which provides the ability to
+cooperatively cancel methods with run to completion semantics also had been
+successfully deployed in these environments.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Syntax and naming questions
+
+This RFC leaves the concrete syntax and naming for the new function and Future
+type open. For the `Future` any name could be used.
+
+For the new async fn syntax, a syntax needs to be chosen that is allowed within
+the Rust 2018 rules. The attribute based differentiation seems to be allowed
+according to the understand of the author, while a new keyword like `completion`
+is problematic.
+
+One early feedback for the RFC was that the attribute could also carry a boolean
+flag on whether completion is supported or not:
+
+```rust
+#[completion(true)]
+async fn runs_to_completion()
+
+#[completion(false)]
+async fn is_synchronously_cancellable()
+```
+
+This might be helpful for code generators as well as the compiler,
+since the completion property mainly becomes a flag which needs to be carried
+forward. Normal (attribute-less) `async fn`s would implicitly gain a
+`#[completion(false)]` attribute. However this shouldn't be a necessarity to
+enable the feature.
+
+Another early feedback was that postfix keywords might be allowed within the
+current grammar, so that `async completion fn` could be an alternative to the
+attribute. This would need to be verified.
+
+## Sub-typing relationships between `Future` and `RunToCompletionFuture`
+
+This RFC leaves it open whether there should be a type relation between
+`Future` and `RunToCompletionFuture`, or whether we just need to teach the
+compiler about both types, and when the `.await` of the other type is possible.
+
+Normal `Future`s can be used whenever a `RunToCompletionFuture` is required.
+In that case the `Future`s support cancellation by drop, but the caller would
+not make use of the capabilty and always drive them to completion.
+
+A blanket impl like the following might be a possibility::
+
+```rust
+impl<F, T> RunToCompletionFuture for F
+where F: Future<Output=T> {
+    type Output = T;
+
+    unsafe fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // forward the poll call to the `Future` implementation
+        <Self as Future>::poll(self, cx)
+    }
+}
+```
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+The RFC outlines a few additional ideas which will be interesting to look at
+in the future:
+
+## Keyword for async run to completion function
+
+A future Edition of Rust could adopt a different keyword for async completion
+functions in order to reduce the verbosity. This should be done
+after carefully studying the usage of the various `async fn` types.
```

Comment:
```suggestion
after carefully studying the usage of the various `async fn` types.

It would also be possible to make `RunToCompletionFuture` be the default.  `#[completion] async` would become just `async`, and `async` would become `#[completion(false)] async` (or similar).  This transformation can be performed automatically.
```

Reply:
It's totally possible. But I would rather not recommend this, since those kinds of semantic changes mean a lot of churn for people updating their projects, as well as might confuse people which work on projects in both editions. I therefore don't expect this to face a lot of acceptance.